File: src/model/boundary.cpp
/// @file boundary.cpp

#include "model/boundary.h"

#include <algorithm>
#include <cmath>
#include <cstdint>
#include <map>
#include <optional>
#include <set>
#include <string>
#include <vector>

#include "util/string_util.h"

namespace coverwise {
namespace model {

using util::IsNumeric;
using util::JsNumberToString;
using util::ToDouble;

namespace {

/// @brief Format an integer value as a string (matches JS String(value)).
std::string FormatInteger(double value) { return JsNumberToString(std::round(value)); }

/// @brief Format a float value as the shortest round-trip string.
///
/// Uses the canonical JS Number-to-String algorithm so newly computed boundary
/// values (min-step, max+step, ...) render identically to the TypeScript port.
std::string FormatFloat(double value) { return JsNumberToString(value); }

}  // namespace

Parameter ExpandBoundaryValues(const Parameter& param, const BoundaryConfig& config) {
  // Generate boundary values.
  std::vector<double> boundary_nums;
  if (config.type == BoundaryConfig::Type::kInteger) {
    double step = 1.0;
    boundary_nums = {
        config.min_value - step, config.min_value, config.min_value + step,
        config.max_value - step, config.max_value, config.max_value + step,
    };
  } else {
    boundary_nums = {
        config.min_value - config.step, config.min_value, config.min_value + config.step,
        config.max_value - config.step, config.max_value, config.max_value + config.step,
    };
  }

  // Map numeric identity to the original value index. Keeping the original
  // spelling (e.g. "1.0") avoids changing seed/constraint identity when the
  // numeric value is deduplicated with a generated boundary.
  std::map<double, std::optional<size_t>> numeric_values;
  std::vector<size_t> non_numeric_indices;
  for (size_t i = 0; i < param.values.size(); ++i) {
    const auto& v = param.values[i];
    const double numeric = IsNumeric(v) ? ToDouble(v) : 0.0;
    if (IsNumeric(v) && std::isfinite(numeric)) {
      numeric_values.emplace(numeric, i);
    } else {
      non_numeric_indices.push_back(i);
    }
  }

  // Add boundary values (dedup).
  for (double bv : boundary_nums) {
    if (std::isfinite(bv)) numeric_values.emplace(bv, std::nullopt);
  }

  std::vector<std::string> expanded_values;
  expanded_values.reserve(numeric_values.size() + non_numeric_indices.size());
  std::vector<bool> expanded_invalid;
  std::vector<std::vector<std::string>> expanded_aliases;
  std::vector<std::string> expanded_classes;
  const bool preserve_invalid = !param.invalid().empty();
  const bool preserve_aliases = !param.all_aliases().empty();
  const bool preserve_classes = !param.equivalence_classes().empty();

  auto append_metadata = [&](std::optional<size_t> original) {
    if (preserve_invalid) {
      expanded_invalid.push_back(original ? param.invalid()[*original] : false);
    }
    if (preserve_aliases) {
      expanded_aliases.push_back(original ? param.all_aliases()[*original]
                                          : std::vector<std::string>{});
    }
    if (preserve_classes) {
      expanded_classes.push_back(original ? param.equivalence_classes()[*original] : "");
    }
  };

  for (const auto& [number, original] : numeric_values) {
    if (original) {
      expanded_values.push_back(param.values[*original]);
    } else if (config.type == BoundaryConfig::Type::kInteger) {
      expanded_values.push_back(FormatInteger(number));
    } else {
      expanded_values.push_back(FormatFloat(number));
    }
    append_metadata(original);
  }

  // Append non-numeric values at the end.
  for (size_t original : non_numeric_indices) {
    expanded_values.push_back(param.values[original]);
    append_metadata(original);
  }

  // Preserve per-value metadata by the original value's numeric/string identity.
  // Newly generated boundary values receive the neutral metadata defaults.
  Parameter result(param.name, std::move(expanded_values));
  if (preserve_invalid) result.set_invalid(std::move(expanded_invalid));
  if (preserve_aliases) result.set_aliases(std::move(expanded_aliases));
  if (preserve_classes) result.set_equivalence_classes(std::move(expanded_classes));
  return result;
}

}  // namespace model
}  // namespace coverwise

File: src/ts/model/boundary.ts
/// Boundary value expansion for numeric parameters.

import { isNumeric, jsNumberToString } from '../util/string_util.js';
import { Parameter } from './parameter.js';

/** Boundary value type: integer or float. */
export enum BoundaryType {
  Integer = 'integer',
  Float = 'float',
}

/** Configuration for boundary value expansion of a numeric parameter. */
export interface BoundaryConfig {
  type: BoundaryType;
  minValue: number;
  maxValue: number;
  /** Step size for float type (default 1.0 for integer). */
  step: number;
}

/** Create a default BoundaryConfig. */
export function createBoundaryConfig(params?: Partial<BoundaryConfig>): BoundaryConfig {
  return {
    type: params?.type ?? BoundaryType.Integer,
    minValue: params?.minValue ?? 0,
    maxValue: params?.maxValue ?? 0,
    step: params?.step ?? 1.0,
  };
}

/** Format an integer value as a string (matches C++ JsNumberToString). */
function formatInteger(value: number): string {
  return jsNumberToString(Math.round(value));
}

/** Format a float value as the shortest round-trip string. */
function formatFloat(value: number): string {
  return jsNumberToString(value);
}

/**
 * Expand a parameter's values to include boundary values.
 *
 * For integer type, adds: min-1, min, min+1, max-1, max, max+1.
 * For float type, adds: min-step, min, min+step, max-step, max, max+step.
 * Merges with existing values (dedup) and sorts numerically.
 *
 * @param param The original parameter.
 * @param config The boundary configuration specifying the range.
 * @returns A new Parameter with expanded values.
 */
export function expandBoundaryValues(param: Parameter, config: BoundaryConfig): Parameter {
  // Generate boundary values.
  let boundaryNums: number[];
  if (config.type === BoundaryType.Integer) {
    const step = 1.0;
    boundaryNums = [
      config.minValue - step,
      config.minValue,
      config.minValue + step,
      config.maxValue - step,
      config.maxValue,
      config.maxValue + step,
    ];
  } else {
    boundaryNums = [
      config.minValue - config.step,
      config.minValue,
      config.minValue + config.step,
      config.maxValue - config.step,
      config.maxValue,
      config.maxValue + config.step,
    ];
  }

  // Keep the original value index for each numeric identity so spelling and
  // per-value metadata survive deduplication with generated boundaries.
  const numericValues = new Map<number, number | null>();
  const nonNumericIndices: number[] = [];
  for (let i = 0; i < param.values.length; ++i) {
    const v = param.values[i];
    const numeric = isNumeric(v) ? Number(v) : Number.NaN;
    if (isNumeric(v) && Number.isFinite(numeric)) {
      if (!numericValues.has(numeric)) {
        numericValues.set(numeric, i);
      }
    } else {
      nonNumericIndices.push(i);
    }
  }

  // Add boundary values (dedup).
  for (const bv of boundaryNums) {
    if (Number.isFinite(bv) && !numericValues.has(bv)) {
      numericValues.set(bv, null);
    }
  }

  // Sort numerically and format.
  const sortedNums = Array.from(numericValues.keys()).sort((a, b) => a - b);

  const expandedValues: string[] = [];
  const expandedInvalid: boolean[] = [];
  const expandedAliases: string[][] = [];
  const expandedClasses: string[] = [];
  const preserveInvalid = param.invalid.length > 0;
  const preserveAliases = param.allAliases.length > 0;
  const preserveClasses = param.equivalenceClasses.length > 0;

  const appendMetadata = (original: number | null): void => {
    if (preserveInvalid) {
      expandedInvalid.push(original === null ? false : param.invalid[original]);
    }
    if (preserveAliases) {
      expandedAliases.push(original === null ? [] : param.allAliases[original]);
    }
    if (preserveClasses) {
      expandedClasses.push(original === null ? '' : param.equivalenceClasses[original]);
    }
  };
  for (const d of sortedNums) {
    const original = numericValues.get(d) ?? null;
    if (original !== null) {
      expandedValues.push(param.values[original]);
    } else if (config.type === BoundaryType.Integer) {
      expandedValues.push(formatInteger(d));
    } else {
      expandedValues.push(formatFloat(d));
    }
    appendMetadata(original);
  }

  // Append non-numeric values at the end.
  for (const original of nonNumericIndices) {
    expandedValues.push(param.values[original]);
    appendMetadata(original);
  }

  const result = new Parameter(param.name, expandedValues);
  if (preserveInvalid) {
    result.setInvalid(expandedInvalid);
  }
  if (preserveAliases) {
    result.setAliases(expandedAliases);
  }
  if (preserveClasses) {
    result.setEquivalenceClasses(expandedClasses);
  }
  return result;
}
